test-suite/js/src/test/scala/org/scalajs/testsuite/jsinterop/AsyncTest.scala

```diff
@@ -221,15 +221,47 @@ class AsyncTest {
 
       var callbackDone = false
 
-      pAssertType.`then`[Unit] { (x: Int) =>
+      pAssertType.`then` { (x: Int) =>
         assertEquals(42, x)
         callbackDone = true
-        (): Unit | js.Thenable[Unit]
+        ()
       }
 
       processQueue()
 
       assertTrue(callbackDone)
     }
   }
+
+  @Test def JSPromiseTypeInference(): Unit = {
+
+    def typed[A](a: => A): Unit = ()
+
+    PromiseMock.withMockedPromise { processQueue =>
+
+      val promiseInt =
+        js.JSPromise[Int]((resolve: js.Function1[Int, _], error: js.Function1[Any, _]) => resolve(42))
+      typed[js.Thenable[Int]](promiseInt)
+
+      val promiseFromThenable =
+        js.JSPromise[Int](
+          (resolve: js.Function1[js.Thenable[Int], _], error: js.Function1[Any, _]) => resolve(promiseInt)
+        )
+      typed[js.Promise[Int]](promiseFromThenable)
+
+      typed[js.Thenable[Int]](promiseInt.`then`((x: Int) => x + 1))
+
+      typed[js.Thenable[Int]](promiseInt.`then`((x: Int) => promiseInt))
```

Comment:
Returning a `Thenable[Int]` in the function supplied to `then` gives a `Thenable[Int]`.